File: algoritmos/SortTileRecursiveAlgorithm.cpp
#include <iostream>
#include <fstream>
#include <istream>
#include <algorithm>
#include <string>
#include <math.h>
#include "SortTileRecursiveAlgorithm.hpp"
#include "../estructuras/Rect.hpp"
#include "../disk_manipulation_functions/binaryFileFunctions.hpp"
#include "../estructuras/RTreeNode.hpp"

/**
 * @brief Constructor de la clase SortTileRecursiveAlgorithm.
 * \param maxNodeCapacity Cantidad máxima de hijos por nodo.
 * \param numberOfRects Cantidad de rectángulos que componen el RTree.
*/
SortTileRecursiveAlgorithm::SortTileRecursiveAlgorithm(unsigned int maxNodeCapacity, unsigned int numberOfRects) {
  this->maxNodeCapacity = maxNodeCapacity;
  this->numberOfRects = numberOfRects;
}


unsigned int SortTileRecursiveAlgorithm::getNumberOfRects() {
  return this->numberOfRects;
}

unsigned int SortTileRecursiveAlgorithm::getMaxNodeCapacity() {
  return this->maxNodeCapacity;
}

/**
 * @brief Getter para el nombre del algoritmo
*/
std::string SortTileRecursiveAlgorithm::getAlgorithmName() {
  return "NearestX";
}

/**
 * @brief Getter del archivo que contiene los nodos internos del RTree.
*/
std::istream* SortTileRecursiveAlgorithm::getTreeFile() {
  return this->treeFile;
}

/**
 * @brief Getter del archivo que contiene las hojas del RTree.
*/
std::istream* SortTileRecursiveAlgorithm::getLeavesFile() {
  return this->leavesFile;
}

void SortTileRecursiveAlgorithm::setTreeFile(std::fstream* fileStream) {
  this->treeFile = fileStream;
}

void SortTileRecursiveAlgorithm::setLeavesFile(std::fstream* fileStream) {
  this->leavesFile = fileStream;
}

/**
 * @brief Método que determina el criterio de ordenamiento de los rectángulos.
 * SortTileRecursive ordena los rectángulos por su centro en el eje X y después en el Y.
*/
bool SortTileRecursiveAlgorithm::orderCriteria(const Rect& rect1, const Rect& rect2) {
  int rect1Center = floor(rect1.x1 + rect1.x2 / 2);
  int rect2Center = floor(rect2.x1 + rect2.x2 / 2);
  return rect1Center < rect2Center;
}

bool SortTileRecursiveAlgorithm::orderCriteriaY(const Rect& rect1, const Rect& rect2) {
  int rect1Center = floor(rect1.y1 + rect1.y2 / 2);
  int rect2Center = floor(rect2.y1 + rect2.y2 / 2);
  return rect1Center < rect2Center;
}

void SortTileRecursiveAlgorithm::preProcess(Rect* rects, unsigned int arraySize) {
  auto orderLambda = [this](const Rect& rect1, const Rect& rect2) { return this->orderCriteria(rect1, rect2); };
  auto orderLambda2 = [this](const Rect& rect1, const Rect& rect2) { return this->orderCriteriaY(rect1, rect2); };
  std::stable_sort(rects, rects + arraySize, orderLambda);
  unsigned int subSetAmount = std::ceil(std::sqrt((double)numberOfRects / maxNodeCapacity));
  unsigned int subSetSize = subSetAmount * this->maxNodeCapacity;
  unsigned int firstSubSetEnd = subSetSize;
  std::stable_sort(rects, rects + firstSubSetEnd, orderLambda2);
  for (int i = 1; i < subSetAmount; i++) {
    unsigned int subsetStart = subSetSize * i + 1;
    unsigned int subsetEnd = std::min(subSetSize * (i+1), numberOfRects);
    std::stable_sort(rects + subsetStart, rects + subsetEnd, orderLambda2);
  }
}

File: rtree/src/algorithms/SortTileRecursiveAlgorithm.java
package algorithms;

import structs.RTreeNode;
import structs.Rectangle;

import java.util.Arrays;

import static java.lang.Math.*;

/**
 * Algoritmo de construcción de RTree que ordena los rectángulos por su centro en X y luego por su centro en Y
 */
public class SortTileRecursiveAlgorithm extends AbstractRTreeAlgorithm {
    public SortTileRecursiveAlgorithm(int numberOfRects, int maxChildrenPerNode) {
        super(numberOfRects, maxChildrenPerNode);
    }

    @Override
    public String getAlgorithmName() {
        return "SortTileRecursive";
    }

    @Override
    public void rectPreProcessing(Rectangle[] rects) {
        Arrays.sort(rects, Rectangle.compareByXCenter());
        int subSetAmount = (int) floor(sqrt((double)numberOfRects / maxChildrenPerNode));
        int subSetSize = subSetAmount * maxChildrenPerNode;
        Arrays.sort(rects, 0, subSetSize, Rectangle.compareByYCenter());
        for (int i = 1; i < subSetAmount; i++) {
            int subsetStart = subSetSize * i + 1;
            int subsetEnd = min(subSetSize * (i+1), numberOfRects);
            Arrays.sort(rects,subsetStart, subsetEnd, Rectangle.compareByYCenter());
        }
    }

    @Override
    public void nodePreProcessing(RTreeNode[] nodes) {
        Arrays.sort(nodes,(RTreeNode node1, RTreeNode node2) -> Rectangle.compareByXCenter().compare(node1.getMBR(), node2.getMBR()));
        int subSetAmount = (int) floor(sqrt((double)nodes.length / maxChildrenPerNode));
        int subSetSize = subSetAmount * maxChildrenPerNode;
        Arrays.sort(nodes, 0, subSetSize, (RTreeNode node1, RTreeNode node2) -> Rectangle.compareByYCenter().compare(node1.getMBR(), node2.getMBR()));
        for (int i = 1; i < subSetAmount; i++) {
            int subsetStart = subSetSize * i + 1;
            int subsetEnd = min(subSetSize * (i+1), nodes.length);
            Arrays.sort(nodes,subsetStart, subsetEnd, (RTreeNode node1, RTreeNode node2) -> Rectangle.compareByYCenter().compare(node1.getMBR(), node2.getMBR()));
        }
    }
}
